lib/web/sign.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package web
+
+import (
+	"bytes"
+	"fmt"
+	"io/fs"
+	"net/http"
+	"time"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/trace"
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/exp/slices"
+
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/lib/client/identityfile"
+	"github.com/gravitational/teleport/lib/httplib"
+	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/srv"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+/* signCertKeyPair returns the necessary files to set up mTLS for other services
+This is the equivalent of running the tctl command
+As an example, requesting:
+POST /webapi/sites/mycluster/sign
+{
+	"hostname": "pg.example.com",
+	"ttl": "2190h",
+	"format": "db"
+}
+
+Should be equivalent to running:
+   tctl auth sign --host=pg.example.com --ttl=2190h --format=db
+
+This endpoint returns a tar.gz compressed archive containing the required files to setup mTLS for the service.
+*/
+func (h *Handler) signCertKeyPair(w http.ResponseWriter, r *http.Request, p httprouter.Params, site reversetunnel.RemoteSite) (interface{}, error) {
+	req := &signCertKeyPairReq{}
+	if err := httplib.ReadJSON(r, &req); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	virtualFS := identityfile.NewInMemoryConfigWriter()
+
+	mTLSReq := srv.GenerateMTLSFilesRequest{
+		ClusterAPI:          h.auth.proxyClient,
+		Principals:          []string{req.Hostname},
+		OutputFormat:        req.Format,
+		OutputCanOverwrite:  true,
+		OutputLocation:      "server",
+		IdentityFileWriter:  virtualFS,
+		TTL:                 req.TTL,
+		HelperMessageWriter: nil,
+	}
+	filesWritten, err := srv.GenerateMTLSFiles(r.Context(), mTLSReq)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	archiveName := fmt.Sprintf("teleport_mTLS_%s.tar.gz", req.Hostname)
+
+	// https://www.postgresql.org/docs/current/libpq-ssl.html
+	// On Unix systems, the permissions on the private key file must disallow any access to world or group;
+	//  achieve this by a command such as chmod 0600 ~/.postgresql/postgresql.key.
+	// Alternatively, the file can be owned by root and have group read access (that is, 0640 permissions).
+	fileMode := fs.FileMode(teleport.FileMaskOwnerOnly) // 0600
```

Comment:
Shouldn't this logic be in the method that generates the certificates? Not sure it belongs in web apiserver.

Reply:
Yes
I ended up removing the explicit `fileMode` here because that's already the default
https://github.com/gravitational/teleport/blob/380f4859870a6afb334ceaa88bdde7d08ce88381/api/identityfile/identityfile.go#L39
But added a couple of comments explaining the reasoning behind using `0600`
If we need to change this value, we must change it here:
https://github.com/gravitational/teleport/blob/380f4859870a6afb334ceaa88bdde7d08ce88381/lib/client/identityfile/identity.go#L242-L257